src/fal/packages/environment.py

```diff
@@ -0,0 +1,157 @@
+            yield dynamic_dependency.name, dynamic_dependency.version
+
+
+HookRunner = Callable[[FalDbt, Path, str, Dict[str, Any], int], int]
+
+
+@contextmanager
+def _clear_on_fail(path: Path) -> Iterator[None]:
+    try:
+        yield
+    except Exception:
+        shutil.rmtree(path)
+        raise
+
+
+class BaseEnvironment:
+    def setup(self) -> ContextManager[HookRunner]:
+        raise NotImplementedError
+
+
+@dataclass(frozen=True)
+class VirtualPythonEnvironment(BaseEnvironment):
+    requirements: List[str] = field(default_factory=list)
+
+    # TODO: This is not intra-process safe, so if you have 2 running
+    # fal processes that tries to create the same virtual environment
+    # *at the same time* it might fail (the behavior is undefined)·
+    #
+    # We might think about introducing a file-lock, which would
+    # allow this scenerio in the future.
+    _VENV_LOCKS: ClassVar[DefaultDict[threading.Lock]] = defaultdict(threading.Lock)
+
+    def __post_init__(self) -> None:
+        self.requirements.extend(
+            f"{key}=={value}" if value else key
+            for key, value in _get_default_requirements()
+        )
+
+    @cached_property
+    def _key(self) -> str:
+        crypt = hashlib.sha256()
+        crypt.update(" ".join(self.requirements).encode())
+        return crypt.hexdigest()
+
+    @contextmanager
+    def setup(self) -> Iterator[HookRunner]:
+        venv_path = self._create_venv()
+        yield partial(self._run_in_venv, venv_path)
+
+    def _create_venv(self) -> Path:
+        from virtualenv import cli_run
+
+        path = _BASE_VENV_DIR / self._key
+        with self._VENV_LOCKS[self._key]:
+            with _clear_on_fail(path):
+                if path.exists():
+                    return path
+
+                print(f"Creating virtual environment at {path}.")
+                cli_run([str(path)])
+
+                print(f"Installing the requirements: {', '.join(self.requirements)}")
+                pip_path = path / "bin" / "pip"
+                subprocess.check_call([pip_path, "install"] + self.requirements)
+
+        return path
+
+    def _run_in_venv(
+        self,
+        venv_path: Path,
+        fal_dbt: FalDbt,
+        hook_path: Path,
+        arguments: Dict[str, Any],
+        bound_model_name: str,
+        run_index: int,
+    ) -> int:
+        python_path = venv_path / "bin" / "python"
+        data = json.dumps(
+            {
+                "path": str(hook_path),
+                "bound_model_name": bound_model_name,
+                "fal_dbt_config": fal_dbt._serialize(),
+                "arguments": arguments,
+                "run_index": run_index,
+            }
+        )
+
+        process = subprocess.Popen(
+            [
+                python_path,
+                _hook_runner_module.__file__,
+                data,
+            ],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            text=True,
+        )
```

Comment:
ATL: why do we do a subprocess instead of running in this same process? it seems we are already in a thread, right?

Reply:
Of course! This makes a lot of sense. Very cool experiments, though.
I agree that it is very complex without really adding substantial value. Not worth doing.

Thanks for the explanation!